File: python/src/contriwork_market_data/_adapter_helpers.py
"""Shared adapter-side helpers.

Internal — adapter authors import from here to keep boilerplate minimal.
"""

from __future__ import annotations

from collections.abc import AsyncIterator, Awaitable, Callable

from .errors import AdapterFeatureNotSupportedError, MissingCredentialsError
from .types import Ticker

__all__ = [
    "resolve_api_key",
    "streaming_not_supported",
]


async def resolve_api_key(
    *,
    adapter_id: str,
    api_key: str | None,
    api_key_provider: Callable[[], Awaitable[str | None]] | None,
    required: bool,
) -> str | None:
    """Lazy-resolve credentials. CONTRACT.md §6 (SCOPE.md §6).

    If both ``api_key`` and ``api_key_provider`` are set, the provider wins
    so callers can rotate keys at runtime (e.g. DB-backed stores). When
    ``required`` is True and neither source yields a non-empty value,
    raise ``MissingCredentialsError`` lazily on first call.
    """
    resolved: str | None
    if api_key_provider is not None:
        resolved = await api_key_provider()
    else:
        resolved = api_key
    if required and not resolved:
        raise MissingCredentialsError(
            f"adapter {adapter_id} requires authentication but no api_key or "
            "api_key_provider resolved a usable value",
            adapter_id=adapter_id,
        )
    return resolved


def streaming_not_supported(adapter_id: str) -> AsyncIterator[Ticker]:
    """Return an async iterator that raises on the first iteration.

    Used by adapters whose ``Capability.supports_native_streaming`` is
    False to satisfy the :class:`Adapter` protocol. The orchestrator never
    invokes this when polling fallback is active; this path exists so
    callers using the adapter directly fail fast with a clear error.
    """

    async def _gen() -> AsyncIterator[Ticker]:
        # An ``async for`` over an empty tuple makes this function an async
        # generator at type-check time without producing any value. The
        # raise below always fires on the first ``__anext__`` call.
        for _ in ():  # pragma: no cover - never iterates
            yield  # type: ignore[misc]
        raise AdapterFeatureNotSupportedError(
            f"adapter {adapter_id} does not implement native streaming; "
            "use polling fallback via MarketDataClient.subscribe_ticker",
            adapter_id=adapter_id,
        )

    return _gen()

File: csharp/src/Contriwork.MarketData/Internal/AdapterHelpers.cs
using System.Runtime.CompilerServices;

namespace Contriwork.MarketData.Internal;

/// <summary>Shared adapter-side helpers — lazy credentials, stub streams.</summary>
internal static class AdapterHelpers
{
    /// <summary>
    /// Lazily resolve a credential. The provider callback wins over the
    /// static key so callers can rotate credentials at runtime. When
    /// <paramref name="required"/> is set and nothing resolves, throws
    /// <see cref="MissingCredentialsException"/>.
    /// </summary>
    /// <param name="adapterId">Adapter id for error context.</param>
    /// <param name="apiKey">Static key, if any.</param>
    /// <param name="apiKeyProvider">Lazy provider, if any.</param>
    /// <param name="required">Whether a credential is mandatory.</param>
    /// <param name="cancellationToken">Cancellation for the provider call.</param>
    /// <returns>The resolved credential, or <c>null</c> when optional and absent.</returns>
    public static async Task<string?> ResolveApiKeyAsync(
        string adapterId,
        string? apiKey,
        Func<CancellationToken, Task<string?>>? apiKeyProvider,
        bool required,
        CancellationToken cancellationToken)
    {
        var resolved = apiKeyProvider is not null
            ? await apiKeyProvider(cancellationToken).ConfigureAwait(false)
            : apiKey;

        if (required && string.IsNullOrEmpty(resolved))
        {
            throw new MissingCredentialsException(
                $"adapter {adapterId} requires authentication but no api key or "
                + "provider resolved a usable value",
                adapterId);
        }

        return resolved;
    }

    /// <summary>
    /// An async stream that throws <see cref="AdapterFeatureNotSupportedException"/>
    /// on first iteration — used by adapters whose
    /// <see cref="Capability.SupportsNativeStreaming"/> is false.
    /// </summary>
    /// <param name="adapterId">Adapter id for error context.</param>
    /// <param name="cancellationToken">Cancellation token (unused; present for signature parity).</param>
    /// <returns>A stream that fails fast.</returns>
    public static async IAsyncEnumerable<Ticker> StreamingNotSupportedAsync(
        string adapterId,
        [EnumeratorCancellation] CancellationToken cancellationToken)
    {
        await Task.CompletedTask.ConfigureAwait(false);
        cancellationToken.ThrowIfCancellationRequested();
        throw new AdapterFeatureNotSupportedException(
            $"adapter {adapterId} does not implement native streaming; "
            + "use polling fallback via MarketDataClient.SubscribeTickerAsync",
            adapterId);
#pragma warning disable CS0162 // Unreachable yield marks the method as an iterator.
        yield break;
#pragma warning restore CS0162
    }
}
